onnxruntime/core/optimizer/graph_optimizer_registry.cc

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/graph_optimizer_registry.h"
+#include "core/optimizer/graph_transformer_utils.h"
+#include "core/optimizer/selection_and_optimization_func.h"
+#include "core/optimizer/qdq_transformer/constant_folding_dq_node.h"
+
+using namespace onnxruntime;
+using namespace ::onnxruntime::common;
+
+namespace onnxruntime {
+
+GraphOptimizerRegistry::GraphOptimizerRegistry() {}
+
+std::optional<SelectionFunc> GraphOptimizerRegistry::GetSelectionFunc(std::string& name) const {
+  auto lookup = transformer_name_to_selection_func_.find(name);
+  if (lookup != transformer_name_to_selection_func_.end()) {
+    return transformer_name_to_selection_func_.at(name);
+  }
+  LOGS(*logger_, WARNING) << "Can't find selection function of " << name;
+  return std::nullopt;
+}
+
+common::Status GraphOptimizerRegistry::Create(
+    const onnxruntime::SessionOptions* sess_options,
+    const onnxruntime::IExecutionProvider* cpu_ep,
```

Comment:
Are there instances where these 2 params are nullptr? If not, take them as references, can store them as pointers, but return them as references from GetCpuEP and GetSessionOptions. The class has an invariant (enforced here) that the values always exist, hence it's safe to return as references from the 'get' methods. 

Side note the 'Reference' in the function name (e.g. GetCpuEpReference) doesn't really make sense to me. Return an actual reference so you don't have the name arguing with the actual type returned.

Reply:
> Are there instances where these 2 params are nullptr?

No, i don't think so. 
I made the 'get' methods return reference to cpu ep or session options.
The singleton of the GraphOptimizerRegistry and it only being instantiated once with cpu ep, sess options and logger, all these can make sure the values (cpu ep, sess options and logger) always exist and invariant.